server/src/main/java/org/opensearch/wlm/tracker/QueryGroupResourceUsage.java

```diff
@@ -0,0 +1,128 @@
+
+/**
+ * This class is used to track query group level resource usage
+ */
+public abstract class QueryGroupResourceUsage {
+    private double currentUsage;
+
+    /**
+     * getter for value field
+     * @return resource usage value
+     */
+    public double getCurrentUsage() {
+        return currentUsage;
+    }
+
+    public void setCurrentUsage(double currentUsage) {
+        this.currentUsage = currentUsage;
+    }
+
+    public static QueryGroupResourceUsage from(ResourceType resourceType) {
+        if (resourceType == ResourceType.CPU) {
+            return new QueryGroupCpuUsage();
+        } else if (resourceType == ResourceType.MEMORY) {
+            return new QueryGroupMemoryUsage();
+        }
+        throw new IllegalArgumentException("Invalid resource type: " + resourceType + ". It is currently not supported in wlm");
+    }
+
+    /**
+     * Determines whether {@link QueryGroup} is breaching its threshold for the resource
+     * @param queryGroup
+     * @return whether the query group is breaching threshold for this resource
+     */
+    public boolean isBreachingThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+        return getCurrentUsage() > getNormalisedThresholdFor(queryGroup, settings);
+    }
+
+    /**
+     * returns the value by which the resource usage is beyond the configured limit for the query group
+     * @param queryGroup instance
+     * @param settings {@link WorkloadManagementSettings} instance
+     * @return the overshooting limit for the resource
+     */
+    public double getReduceByFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+        return getCurrentUsage() - getNormalisedThresholdFor(queryGroup, settings);
+    }
+
+    /**
+     * initialises the member variable currentUsage
+     * @param tasks list of tasks in the query group
+     * @param timeSupplier nano time supplier
+     */
+    public void initialise(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier) {
+        this.setCurrentUsage(this.calculateResourceUsage(tasks, timeSupplier));
+    }
+
+    /**
+     * normalises configured value with respect to node level cancellation thresholds
+     * @param queryGroup instance
+     * @param settings {@link WorkloadManagementSettings} instance
+     * @return normalised value with respect to node level cancellation thresholds
+     */
+    public abstract double getNormalisedThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings);
+
+    /**
+     * calculates the current resource usage for the query group
+     * @param tasks list of tasks in the query group
+     * @param timeSupplier nano time supplier
+     */
+    protected abstract double calculateResourceUsage(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier);
+
+    /**
+     * class to store cpu usage for the query group
+     */
+    public static class QueryGroupCpuUsage extends QueryGroupResourceUsage {
```

Comment:
When will this object be created? It seems like we have a plan to create this usage object every time the threshold changes for a particular queryGroup which looks unnecessary and creates overhead.

Ideally when a queryGroupId is created, a corresponding QueryGroup is created. Which internally holds a reference to this resourceUsage object which is created only once. Any time we change the threshold, we just keep updating the `currentUsage` via desired mechanism.

Reply:
No this is the point in time resource usage value for the query group. I agree that all we need is refreshing the values for the resource usages for a query group but each refresh ultimately result in updating majority of the objects in `QueryGroupLevelResourceUsageView`. So I think this should not incur any major overhead

Reply:
I have changed the resource usage type to a double so I think this will not be needed.